File: tests/src/main/java/com/github/klee0kai/test/di/base_forest/ForestComponent.java
package com.github.klee0kai.test.di.base_forest;

import com.github.klee0kai.stone.annotations.component.*;
import com.github.klee0kai.stone.lifecycle.StoneLifeCycleOwner;
import com.github.klee0kai.test.mowgli.School;
import com.github.klee0kai.test.mowgli.animal.Horse;
import com.github.klee0kai.test.mowgli.animal.Mowgli;
import com.github.klee0kai.test.mowgli.animal.Snake;

@Component
public interface ForestComponent {

    UnitedModule united();

    IdentityModule identity();

    void inject(Horse horse, StoneLifeCycleOwner stoneLifeCycleOwner);

    void inject(Horse horse);


    void inject(Mowgli mowgli);

    void inject(Snake snake);

    void inject(School school);

    @Init
    void initUnitedModule(UnitedModule unitedModule);

    @Init
    void initUnitedModule(Class<? extends UnitedModule> unitedModule);

    @Init
    void iniAllModules(UnitedModule unitedModule, IdentityModule identityModule);

    @RunGc
    @GcAllScope
    void gcAll();

    @ProtectInjected(timeMillis = 30)
    void protectInjected(Horse horse);


    @ProtectInjected(timeMillis = 30)
    void protectInjected(Mowgli horse);

    @ProtectInjected(timeMillis = 30)
    void protectInjected(School school);

}

File: tests/src/main/java/com/github/klee0kai/test/di/base_forest/ForestComponent.kt
package com.github.klee0kai.test.di.base_forest

import com.github.klee0kai.stone.annotations.component.*
import com.github.klee0kai.stone.lifecycle.StoneLifeCycleOwner
import com.github.klee0kai.test.mowgli.School
import com.github.klee0kai.test.mowgli.animal.Horse
import com.github.klee0kai.test.mowgli.animal.Mowgli
import com.github.klee0kai.test.mowgli.animal.Snake

@Component
interface ForestComponent {
    fun united(): UnitedModule?

    fun identity(): IdentityModule?

    fun inject(horse: Horse?, stoneLifeCycleOwner: StoneLifeCycleOwner?)

    fun inject(horse: Horse?)


    fun inject(mowgli: Mowgli?)

    fun inject(snake: Snake?)

    fun inject(school: School?)

    @Init
    fun initUnitedModule(unitedModule: UnitedModule?)

    @Init
    fun iniAllModules(unitedModule: UnitedModule?, identityModule: IdentityModule?)

    @RunGc
    @GcAllScope
    fun gcAll()

    @ProtectInjected(timeMillis = 30)
    fun protectInjected(horse: Horse?)


    @ProtectInjected(timeMillis = 30)
    fun protectInjected(horse: Mowgli?)

    @ProtectInjected(timeMillis = 30)
    fun protectInjected(school: School?)
}
